File: Chapter 03/C#/Section_3_4_OrderProcessor_Refactor/Before/OrderProcessor.cs
using System;

namespace Chapter03.OrderProcessorRefactor.Before;

/// <summary>
/// ANTI-PATTERN: The "God Class" / Monolithic Transaction Script.
/// 
/// ARCHITECTURE PROBLEM: This class is a massive violation of the Single 
/// Responsibility Principle (SRP) and is tightly coupled to four distinct domains.
/// 
/// If this were a real enterprise application, this one method would:
/// 1. Contain complex business validation rules.
/// 2. Talk to a third-party Payment Gateway API (like Stripe).
/// 3. Connect to a SQL Database to update inventory tables.
/// 4. Talk to a third-party Email API (like SendGrid).
/// 
/// WHY THIS FAILS:
/// - Testing: You cannot test the validation logic without accidentally triggering 
///   real payment calls or database updates.
/// - Maintenance: Four different teams (Billing, Inventory, Communications, Sales) 
///   will constantly be modifying this exact same file, causing merge conflicts.
/// - Brittleness: A timeout error from the Email API could cause the entire 
///   payment process to roll back or crash!
/// </summary>
public class OrderProcessor
{
    public string Process(Order order)
    {
        // 🚨 ARCHITECTURE WARNING: Domain Rule Violation
        // 1. Validation
        Console.WriteLine("  [Validate] Validating order...");
        if (order.Items.Count == 0 || order.Total <= 0)
        {
            throw new InvalidOperationException("Order is invalid.");
        }

        // 🚨 ARCHITECTURE WARNING: External Integration Coupling
        // 2. Payment Processing
        Console.WriteLine($"  [Payment] Processing payment for ${order.Total}...");
        bool paymentSuccess = true; 

        // 🚨 ARCHITECTURE WARNING: Infrastructure & Communications Coupling
        // 3. Inventory Update & 4. Confirmation Email
        if (paymentSuccess)
        {
            Console.WriteLine("  [Inventory] Updating inventory...");
            Console.WriteLine($"  [Notify] Sending confirmation email to {order.CustomerEmail}...");
            return "Order processed successfully.";
        }
        else
        {
            return "Payment failed.";
        }
    }
}

File: Chapter 03/JS/section_3_4_orderProcessor_refactor/before/orderProcessor.js
/**
 * ANTI-PATTERN: The "God Class" / Monolithic Transaction Script.
 * * ARCHITECTURE PROBLEM: This class is a "Swiss Army Knife" that knows too 
 * much. It violates the Single Responsibility Principle (SRP) by managing 
 * four distinct domains: Validation, Payments, Inventory, and Notifications.
 * * WHY THIS IS DANGEROUS:
 * - Fragility: If the Email API is down, the whole order processing fails.
 * - Testing: You cannot test the validation logic without triggering 
 * console logs (or real side effects) for payments and inventory.
 * - Rigidity: Changing your payment provider (e.g., from Stripe to PayPal) 
 * requires you to open and modify this core business logic file.
 */
class OrderProcessor {
    /**
     * Processes an order by executing four different responsibilities in a row.
     * @param {Order} order - The order to be processed.
     */
    process(order) {
        
        // 🚨 ARCHITECTURE WARNING: Domain Rule Violation
        // Responsibility 1: Validation. 
        // This should be in a separate validator.
        console.log("  [Validate] Validating order...");
        if (order.items.length === 0 || order.total <= 0) {
            throw new Error("Order is invalid.");
        }

        // 🚨 ARCHITECTURE WARNING: External Integration Coupling
        // Responsibility 2: Payment Processing. 
        // This couples the business flow to a specific billing implementation.
        console.log(`  [Payment] Processing payment for $${order.total.toFixed(2)}...`);
        const paymentSuccess = true;

        if (paymentSuccess) {
            // 🚨 ARCHITECTURE WARNING: Infrastructure Coupling
            // Responsibility 3: Inventory Management.
            console.log("  [Inventory] Updating inventory...");

            // 🚨 ARCHITECTURE WARNING: Communications Coupling
            // Responsibility 4: Notifications.
            console.log(`  [Notify] Sending confirmation email to ${order.customerEmail}...`);
            
            return "Order processed successfully.";
        } else {
            return "Payment failed.";
        }
    }
}

module.exports = OrderProcessor;